src/app/public/modules/datepicker/datepicker-input.directive.ts

```diff
@@ -64,230 +69,332 @@ const SKY_DATEPICKER_VALIDATOR = {
+    if (changes.minDate) {
+      this.onValidatorChange();
+      this.datepickerComponent.minDate = this.minDate;
     }
 
-    if (changes['maxDate']) {
-      this._validatorChange();
-      this.skyDatepickerInput.setMaxDate(this.maxDate);
+    if (changes.maxDate) {
+      this.onValidatorChange();
+      this.datepickerComponent.maxDate = this.maxDate;
     }
 
-    if (changes['startingDay']) {
-      this._validatorChange();
-      this.skyDatepickerInput.startingDay = this.startingDay;
+    if (changes.startingDay) {
+      this.onValidatorChange();
+      this.datepickerComponent.startingDay = this.startingDay;
     }
   }
 
+  public ngOnDestroy(): void {
+    this.ngUnsubscribe.next();
+    this.ngUnsubscribe.complete();
+  }
+
   @HostListener('change', ['$event'])
-  public onChange(event: any) {
-    this.writeValue(event.target.value);
+  public onInputChange(event: any) {
+    this.isFirstChange = false;
+    this.value = event.target.value;
   }
 
   @HostListener('blur')
-  public onBlur() {
-    this._onTouched();
+  public onInputBlur(): void {
+    this.onTouched();
   }
 
-  public registerOnChange(fn: (value: any) => any): void { this._onChange = fn; }
-
-  public registerOnTouched(fn: () => any): void { this._onTouched = fn; }
-
-  public registerOnValidatorChange(fn: () => void): void { this._validatorChange = fn; }
-
-  public setDisabledState(isDisabled: boolean) {
-    this.disabled = isDisabled;
+  @HostListener('keyup')
+  public onInputKeyup(): void {
+    this.control.markAsDirty();
   }
 
-  public writeValue(value: any) {
-    this.modelValue = value;
+  public writeValue(value: any): void {
+    this.value = value;
   }
 
-  public validate(control: AbstractControl): {[key: string]: any} {
-    let value = control.value;
+  public validate(control: AbstractControl): ValidationErrors {
+    if (!this.control) {
+      this.control = control;
+    }
+
+    if (this.skyDatepickerNoValidate) {
+      return;
+    }
+
+    const value: any = control.value;
 
     if (!value) {
-      return undefined;
+      return;
     }
 
-    let dateValue = this.dateFormatter.getDateFromString(value, this.dateFormat);
+    const dateValue = this.getDateValue(value);
+    const isDateValid = (dateValue && this.dateFormatter.dateIsValid(dateValue));
+
+    if (!isDateValid) {
+      // Mark the invalid control as touched so that the input's invalid CSS styles appear.
+      // (This is only required when the invalid value is set by the FormControl constructor.)
+      this.control.markAsTouched();
 
-    if (!this.dateFormatter.dateIsValid(dateValue) && !this.skyDatepickerNoValidate) {
       return {
         'skyDate': {
-          invalid: control.value
+          invalid: value
         }
       };
     }
 
-    if (this.minDate &&
-      this.dateFormatter.dateIsValid(this.minDate) &&
-      this.dateFormatter.dateIsValid(value) &&
-      value < this.minDate) {
+    const minDate = this.minDate;
```

Comment:
Personally, it creates a cleaner implementation of the fairly complex conditional in the next few lines. If we remove the `this.` it's easier to follow the logic trail.

Also, in JavaScript, it takes time to do a property lookup in a large object (e.g., `this`). By setting it to a local constant, the lookup only occurs once, instead of three times in the condition.

Reply:
:+1: